### PR TITLE
More CMake refactoring and touch-ups

### PR DESCRIPTION
**Pull request by lgritz**
* Eschew directory-level include_directories and add_definitions in favor
  of explicitly tying them to targets and using PRIVATE/PUBLIC.

* Use standard idiom CMAKE_CXX_STANDARD instead of custom USE_CPP.

### CODE REVIEW

File: src/cmake/compiler.cmake [hunk 1]
```diff
@@ -195,17 +200,7 @@ endif ()
...
-        set (CSTD_FLAGS "-std=c++17")
-    elseif (USE_CPP VERSION_EQUAL 14)
-        message (STATUS "Building for C++14")
-        set (CSTD_FLAGS "-std=c++14")
-    else ()
-        message (STATUS "Building for C++11")
-        set (CSTD_FLAGS "-std=c++11")
-    endif ()
-    add_compile_options (${CSTD_FLAGS})
+    set (CSTD_FLAGS "-std=c++${CMAKE_CXX_STANDARD}")
```
[the remark this answers is not in the document]
Comment (lgritz) [reply src/cmake/compiler.cmake]: I believe you might be right. I was being conservative by passing it to check_cxx_source_runs that I use to check whether we can use std::regex or boost::regex. But looking at the docs, I think the ordinary flags are passed there automatically. I'll double check.

**Comment by lgritz** [comment]
BTW, you can see that what I'm doing is ever few days I spend an evening reading a chapter of the Professional CMake book and doing the fixes from whatever I learn from that chapter. Bit by bit it's getting cleaned up as I repay 10 years of technical debt on the build system. :-)

File: src/cmake/compiler.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (lgritz) [reply src/cmake/compiler.cmake]: Yeah, looks safe. I made the change. We'll see if it passes the CI everywhere now.

[the remark this answers is not in the document]
Comment (lgritz) [reply src/cmake/compiler.cmake]: Yep, seems fine, passes all tests.

**Comment by fpsunflower** [comment]
LGTM!

**Pull request merged**